capsules/src/print_log.rs

```diff
@@ -0,0 +1,276 @@
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::PrintLog as usize;
+
+/// Ids for read-only allow buffers
+mod ro_allow {
+    /// Before the allow syscall was handled by the kernel,
+    /// console used allow number "1", so to preserve compatibility
+    /// we still use allow number 1 now.
+    pub const WRITE: usize = 1;
+    /// The number of allow buffers the kernel stores for this grant
+    pub const COUNT: u8 = 2;
+}
+
+/// Ids for read-write allow buffers
+mod rw_allow {
+    /// Before the allow syscall was handled by the kernel,
+    /// console used allow number "1", so to preserve compatibility
+    /// we still use allow number 1 now.
+    pub const _READ: usize = 0;
+    /// The number of allow buffers the kernel stores for this grant
+    pub const COUNT: u8 = 0;
+}
+
+#[derive(Default)]
+pub struct App {
+    write_len: usize,    // Length of write
+    writing: bool,       // Are we in the midst of a write
+    pending_write: bool, // Are we waiting to write
+    tx_counter: usize,   // Used to keep order of writes
+}
+
+pub struct PrintLog<'a, A: Alarm<'a>> {
+    apps: Grant<
+        App,
+        UpcallCount<3>,
+        AllowRoCount<{ ro_allow::COUNT }>,
+        AllowRwCount<{ rw_allow::COUNT }>,
+    >,
+    tx_in_progress: Cell<bool>, // If true there's an ongoing write so others must wait
+    tx_counter: Cell<usize>,    // Sequence number for writes from different processes
+    alarm: &'a A,               // Timer for trying to send  more
+}
+
+impl<'a, A: Alarm<'a>> PrintLog<'a, A> {
+    pub fn new(
+        alarm: &'a A,
+        grant: Grant<
+            App,
+            UpcallCount<3>,
+            AllowRoCount<{ ro_allow::COUNT }>,
+            AllowRwCount<{ rw_allow::COUNT }>,
+        >,
+    ) -> PrintLog<'a, A> {
+        PrintLog {
+            apps: grant,
+            tx_in_progress: Cell::new(false),
+            tx_counter: Cell::new(0),
+            alarm: alarm,
+        }
+    }
+
+    /// Internal helper function for starting up a new print; allocate a sequence number and
+    /// start the send state machine.
+    fn send_new(
+        &self,
+        app: &mut App,
+        kernel_data: &GrantKernelData,
+        len: usize,
+    ) -> Result<(), ErrorCode> {
+        // We are already writing
+        if app.writing || app.pending_write {
+            return Err(ErrorCode::BUSY);
+        }
+        app.write_len = kernel_data
+            .get_readonly_processbuffer(ro_allow::WRITE)
+            .map_or(0, |write| write.len())
+            .min(len);
+        // We have nothing to write
+        if app.write_len == 0 {
+            return Err(ErrorCode::NOMEM);
+        }
+        // Order the prints through a global counter.
+        app.tx_counter = self.tx_counter.get();
+        self.tx_counter.set(app.tx_counter.wrapping_add(1));
```

Comment:
When this wraps, an application with u32::MAX as its sequence number could get starved indefinitely. Obviously that is a lot of prints so maybe not worth worrying about?

Reply:
Ah, good point; once you wrap around, all of the outstanding writes from before the wraparound could be reordered, or starved indefinitely (if the console is always fully busy). I hadn't considered this case. As soon as it falls, idle, starvation isn't a problem, but re-ordering will be in all cases. So I think this is a major concern. I'll work on a fix.

Reply:
I don't think that's right; I handled wraparound. Sorry, this code was a month old so I forgot.

https://github.com/tock/tock/blob/5d939e1363d41e7fda2167581b7cc69f706a6b92/capsules/src/print_log.rs#L207

For example, suppose for simplicity that it's using an 8-bit rather than 32-bit counter. There are four writes enqueued: 0xfd , 0xfe, 0xff and 0x00. 0xfd is currently in flight.

When the alarm fires, 

https://github.com/tock/tock/blob/5d939e1363d41e7fda2167581b7cc69f706a6b92/capsules/src/print_log.rs#L179

after checking if it should issue a callback, the handler goes to figure out which write to write next. `self.tx_counter = 1` since there is write with 0x00 enqueued.

Suppose that the order of the writes in `apps.iter()` is 0, 0xfe, 0xff, 0xfd. The loop

https://github.com/tock/tock/blob/5d939e1363d41e7fda2167581b7cc69f706a6b92/capsules/src/print_log.rs#L201

Here are the iterations:
  1. check if `1.wrapping_sub(0) < (0xff / 2)` so `1 < 0x7f`. It is,  so `seqno = 0`.
  2. check if `0.wrapping_sub(0xfe) < (0xff / 2)`, so `2 < 0x7f`. It is, so `seqno = 0xfe`.
  3. check if `0xfe.wrapping_sub(0xff) < (0xff / 2)`, so `0xff < 0x7f`. It isn't, so `seqno = 0xfe`.
  4. check if `0xfe.wrapping_sub(0xfd) < (0xff / 2), so `0x01 < 0x7f`. It is, so `seqno = 0xfd`.

Since the gap between the lowest and the highest sequence numbers is at most the number of processes, this works as long as the number of processes is less than `usize::MAX / 2`. The sequence number increments only if there's a valid grant, and the send can only fail if there isn't a valid grant. So, there isn't a danger that an application calls this in a loop, there's a problem with its grant and it causes the sequence numbers to advance and spread more than the number of processes.

Reply:
Ok, I agree that wraparound is handled -- sorry for making you revisit this, I should have checked the actual code. I was thrown off by this line in your PR description: "While a client (process) finishes sending, the driver looks for the client with a write pending that has the lowest sequence number.", and did not verify that the actual implementation wasn't in fact smarter.